### PR TITLE
[Feat] custom ring all-reduce

### PR DESCRIPTION
**Pull request by Jackmin801**
`torchrun --nproc-per-node 2 scripts/all_reduce_test.py --backend custom --transfer_dtype uint8` on `saturn`

| Backend | Dtype    | Time Taken (s) | Bandwidth (GB/s) |
|---------|----------|----------------|------------------|
| gloo    | float32  | 3.12           | 1.28             |
| custom  | float32  | 6.90           | 0.58             |
| custom  | bfloat16 | 4.73           | 0.85             |
| custom  | uint8    | 8.99          | 0.45             |

However under the 500mbps test, it beats the gloo one

image

1B all reduce takes 102s under 500mbps conditions (I think the 4 procs share the limit tho so maybe we can divide the time taken by 4 if the links are all individually 500mbps)
image

### CODE REVIEW

File: src/zeroband/compression.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import torch
+import numpy as np
+from typing import Tuple
+import math
+from concurrent.futures import ThreadPoolExecutor
+import os
+
```
Comment (samsja) [inline src/zeroband/compression.py:11]: just had that this was inspired by hivemind for giving proper credits :)

File: scripts/all_reduce_test.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+def main(config: Config):
+    world_info = get_world_info()
+
+    mat = torch.rand(1, config.size_model)
+
+    logger.info(
+        f"\n ======== Benchmark all reduce between {world_info.world_size} gpus over {world_info.nnodes} nodes =========\n"
+    )
+
+    all_reduce = ALL_REDUCE_FN[config.backend]
```
Comment (samsja) [inline scripts/all_reduce_test.py:45]: can you implement the ability to use te 8bit quant here ?

Ideally we would just use a big Enum for "fp32, fp16, qint8, ... "

**Review by samsja — COMMENTED** [review]
lgtm added some small non blocking comments

File: src/zeroband/compression.py (hunk 1 above)
Comment (Jackmin801) [reply src/zeroband/compression.py:11]: yea sure :)

File: scripts/all_reduce_test.py (hunk 2 above)
Comment (Jackmin801) [reply scripts/all_reduce_test.py:45]: yeep done

File: src/zeroband/C/compression.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+from typing import Tuple
+import torch
+from torch.utils.cpp_extension import load
+from pathlib import Path
+
+COMPRESS_CSRC_PATH = Path(__file__).parent.parent / "csrc" / "compress.cpp"
```
Comment (samsja) [inline src/zeroband/C/compression.py]: you seems to not use this env var

Comment (Jackmin801) [reply src/zeroband/C/compression.py]: oh my bad haha. nice catch!

**Review by samsja — APPROVED** [review]
lfgtm

**Pull request merged**